File: examples/desktop/python/src/pomodoro/pomodoro.py
"""Pomodoro timer state machine and session store."""

from __future__ import annotations

import json
import os
import shutil
from datetime import datetime, timezone
from enum import Enum
from pathlib import Path
from typing import Any


class Phase(Enum):
    IDLE = "idle"
    WORKING = "working"
    SHORT_BREAK = "short_break"
    LONG_BREAK = "long_break"


_DEFAULT_DIR = Path.home() / ".pomodoro"
_DEFAULT_FILE = _DEFAULT_DIR / "sessions.json"
_SEED_FILE = Path(__file__).resolve().parent.parent.parent / "seed.json"


class PomodoroTimer:
    """Core timer state machine with session persistence."""

    def __init__(self) -> None:
        self.phase = Phase.IDLE
        self.paused = False
        self.time_remaining_sec = 0
        self.time_elapsed_sec = 0
        self.current_tag: str | None = None
        self.cycle_count = 0  # pomodoros completed in current cycle (resets after long break)
        self.sessions: list[dict[str, Any]] = []
        self.settings: dict[str, Any] = {
            "work_duration_sec": 1500,
            "short_break_sec": 300,
            "long_break_sec": 900,
            "long_break_interval": 4,
        }
        self._current_session_start: str | None = None
        self._file_path: Path | None = None

    # --- Persistence ---

    def _resolve_path(self) -> Path:
        env = os.environ.get("POMODORO_FILE")
        if env:
            return Path(env).expanduser().resolve()
        return _DEFAULT_FILE

    def load(self) -> None:
        """Load sessions from disk, seeding from seed.json if missing."""
        path = self._resolve_path()
        self._file_path = path
        if not path.exists():
            path.parent.mkdir(parents=True, exist_ok=True)
            if _SEED_FILE.exists():
                shutil.copy2(_SEED_FILE, path)
            else:
                path.write_text(json.dumps({"sessions": [], "settings": self.settings}, indent=2))
        data = json.loads(path.read_text())
        self.sessions = data.get("sessions", [])
        stored_settings = data.get("settings")
        if stored_settings:
            self.settings.update(stored_settings)

    def save(self) -> None:
        """Write sessions and settings to disk."""
        path = self._file_path or self._resolve_path()
        path.parent.mkdir(parents=True, exist_ok=True)
        data = {"sessions": self.sessions, "settings": self.settings}
        path.write_text(json.dumps(data, indent=2) + "\n")

    # --- Timer actions ---

    def start(self, tag: str | None = None) -> dict[str, Any]:
        """Start a work session."""
        if self.phase != Phase.IDLE:
            return {"error": "Timer is not idle"}
        self.phase = Phase.WORKING
        self.paused = False
        self.time_remaining_sec = self.settings["work_duration_sec"]
        self.time_elapsed_sec = 0
        self.current_tag = tag
        self._current_session_start = datetime.now(timezone.utc).isoformat()
        return {"phase": self.phase.value, "tag": tag}

    def pause(self) -> dict[str, Any]:
        """Pause the timer."""
        if self.phase == Phase.IDLE or self.paused:
            return {"error": "Cannot pause"}
        self.paused = True
        return {"phase": self.phase.value, "paused": True}

    def resume(self) -> dict[str, Any]:
        """Resume the timer."""
        if not self.paused:
            return {"error": "Not paused"}
        self.paused = False
        return {"phase": self.phase.value, "paused": False}

    def skip(self) -> dict[str, Any]:
        """Skip to next phase."""
        if self.phase == Phase.IDLE:
            return {"error": "Nothing to skip"}
        if self.phase == Phase.WORKING:
            # Record completed session
            self._record_session(completed=True)
            self.cycle_count += 1
            # Transition to break
            if self.cycle_count % self.settings["long_break_interval"] == 0:
                self._enter_phase(Phase.LONG_BREAK)
            else:
                self._enter_phase(Phase.SHORT_BREAK)
        else:
            # On a break — go to idle
            self._enter_phase(Phase.IDLE)
        return {"phase": self.phase.value}

    def stop(self) -> dict[str, Any]:
        """Abandon current session, return to idle."""
        if self.phase == Phase.IDLE:
            return {"error": "Already idle"}
        if self.phase == Phase.WORKING:
            self._record_session(completed=False)
        self._enter_phase(Phase.IDLE)
        self.cycle_count = 0
        return {"phase": "idle"}

    def set_tag(self, label: str) -> dict[str, Any]:
        """Set or change tag on the current working session."""
        if self.phase != Phase.WORKING:
            return {"error": "Not in a working session"}
        self.current_tag = label
        return {"tag": label}

    def tag_session(self, session_id: str, label: str) -> dict[str, Any]:
        """Re-tag a completed session."""
        for s in self.sessions:
            if s["id"] == session_id:
                s["tag"] = label
                self.save()
                return {"id": session_id, "tag": label}
        return {"error": "Session not found"}

    def delete_session(self, session_id: str) -> dict[str, Any]:
        """Delete a session."""
        for s in self.sessions:
            if s["id"] == session_id:
                self.sessions.remove(s)
                self.save()
                return {"id": session_id}
        return {"error": "Session not found"}

    # --- Tick ---

    def tick(self) -> bool:
        """Called every 1 second. Returns True if a phase transition occurred."""
        if self.phase == Phase.IDLE or self.paused:
            return False

        self.time_remaining_sec -= 1
        self.time_elapsed_sec += 1

        if self.time_remaining_sec <= 0:
            self.time_remaining_sec = 0
            return self._handle_phase_end()
        return False

    # --- Helpers ---

    def next_id(self) -> str:
        """Generate next session ID (s-N)."""
        max_num = 0
        for s in self.sessions:
            sid = s.get("id", "")
            if sid.startswith("s-"):
                try:
                    max_num = max(max_num, int(sid[2:]))
                except ValueError:
                    pass
        return f"s-{max_num + 1}"

    def _record_session(self, completed: bool) -> None:
        """Record the current working session."""
        now = datetime.now(timezone.utc).isoformat()
        session: dict[str, Any] = {
            "id": self.next_id(),
            "tag": self.current_tag or "Untitled",
            "category": "work",
            "started_at": self._current_session_start or now,
            "ended_at": now,
            "duration_sec": self.time_elapsed_sec,
            "completed": completed,
        }
        self.sessions.append(session)
        self.save()

    def _enter_phase(self, phase: Phase) -> None:
        """Transition to a new phase."""
        self.phase = phase
        self.paused = False
        self.current_tag = None
        self._current_session_start = None

        if phase == Phase.WORKING:
            self.time_remaining_sec = self.settings["work_duration_sec"]
            self.time_elapsed_sec = 0
        elif phase == Phase.SHORT_BREAK:
            self.time_remaining_sec = self.settings["short_break_sec"]
            self.time_elapsed_sec = 0
        elif phase == Phase.LONG_BREAK:
            self.time_remaining_sec = self.settings["long_break_sec"]
            self.time_elapsed_sec = 0
        elif phase == Phase.IDLE:
            self.time_remaining_sec = 0
            self.time_elapsed_sec = 0

    def _handle_phase_end(self) -> bool:
        """Handle timer reaching zero. Returns True (phase transition occurred)."""
        if self.phase == Phase.WORKING:
            self._record_session(completed=True)
            self.cycle_count += 1
            if self.cycle_count % self.settings["long_break_interval"] == 0:
                self._enter_phase(Phase.LONG_BREAK)
            else:
                self._enter_phase(Phase.SHORT_BREAK)
        elif self.phase in (Phase.SHORT_BREAK, Phase.LONG_BREAK):
            self._enter_phase(Phase.IDLE)
        return True

    # --- Stats ---

    def today_sessions(self) -> list[dict[str, Any]]:
        """Return sessions from today, most recent first."""
        today_str = datetime.now(timezone.utc).strftime("%Y-%m-%d")
        today = [
            s for s in self.sessions
            if s.get("started_at", "").startswith(today_str)
        ]
        today.sort(key=lambda s: s.get("started_at", ""), reverse=True)
        return today

    def today_stats(self) -> dict[str, Any]:
        """Compute today's statistics."""
        today = self.today_sessions()
        completed = [s for s in today if s.get("completed")]
        total_focus_sec = sum(s.get("duration_sec", 0) for s in completed)

        # Streak: count consecutive days with at least one completed session
        from datetime import timedelta
        streak = 0
        best_streak = 0
        check_date = datetime.now(timezone.utc).date()
        all_dates = set()
        for s in self.sessions:
            if s.get("completed"):
                try:
                    dt = datetime.fromisoformat(s["started_at"].replace("Z", "+00:00"))
                    all_dates.add(dt.date())
                except (ValueError, KeyError):
                    pass

        while check_date in all_dates:
            streak += 1
            check_date -= timedelta(days=1)

        # Best streak (simple scan)
        if all_dates:
            sorted_dates = sorted(all_dates)
            current_run = 1
            best_streak = 1
            for i in range(1, len(sorted_dates)):
                if (sorted_dates[i] - sorted_dates[i - 1]).days == 1:
                    current_run += 1
                    best_streak = max(best_streak, current_run)
                else:
                    current_run = 1

        return {
            "today_completed": len(completed),
            "today_total_focus_min": total_focus_sec // 60,
            "streak_days": streak,
            "best_streak_days": best_streak,
        }

    def session_salience(self, session: dict[str, Any]) -> tuple[float, str]:
        """Compute salience and reason for a session based on age."""
        ended_at = session.get("ended_at", "")
        if not ended_at:
            return 0.2, "No end time"
        try:
            ended = datetime.fromisoformat(ended_at.replace("Z", "+00:00"))
        except ValueError:
            return 0.2, "Unknown"
        now = datetime.now(timezone.utc)
        diff = now - ended
        minutes = diff.total_seconds() / 60
        hours = diff.total_seconds() / 3600

        if minutes < 60:
            return 0.6, f"Completed {int(minutes)} min ago"
        elif hours < 3:
            return 0.4, f"Completed {int(hours)}h ago"
        else:
            return 0.2, f"Completed {int(hours)}h ago"

    def format_time(self, seconds: int) -> str:
        """Format seconds as MM:SS."""
        m, s = divmod(max(0, seconds), 60)
        return f"{m:02d}:{s:02d}"

    def description(self) -> str:
        """Human description of current state for discovery."""
        if self.phase == Phase.IDLE:
            stats = self.today_stats()
            return f"Pomodoro timer: idle, {stats['today_completed']} sessions today"
        elif self.phase == Phase.WORKING:
            tag_part = f" on '{self.current_tag}'" if self.current_tag else ""
            return f"Working: {self.format_time(self.time_remaining_sec)} remaining{tag_part}"
        elif self.phase == Phase.SHORT_BREAK:
            return f"Short break: {self.format_time(self.time_remaining_sec)} remaining"
        elif self.phase == Phase.LONG_BREAK:
            return f"Long break: {self.format_time(self.time_remaining_sec)} remaining"
        return "Pomodoro timer"

File: examples/desktop/rust/src-tauri/src/pomodoro.rs
use std::fs;
use std::path::PathBuf;

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

/// Timer phase.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Phase {
    Idle,
    Working,
    ShortBreak,
    LongBreak,
}

impl Phase {
    pub fn as_str(&self) -> &'static str {
        match self {
            Phase::Idle => "idle",
            Phase::Working => "working",
            Phase::ShortBreak => "short_break",
            Phase::LongBreak => "long_break",
        }
    }
}

/// A completed (or abandoned) pomodoro session.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Session {
    pub id: String,
    pub tag: String,
    pub category: String,
    pub started_at: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub ended_at: Option<String>,
    pub duration_sec: u64,
    pub completed: bool,
}

/// Persisted data file structure.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SessionFile {
    pub sessions: Vec<Session>,
    pub settings: Settings,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Settings {
    pub work_duration_sec: u64,
    pub short_break_sec: u64,
    pub long_break_sec: u64,
    pub long_break_interval: u32,
}

impl Default for Settings {
    fn default() -> Self {
        Self {
            work_duration_sec: 1500,
            short_break_sec: 300,
            long_break_sec: 900,
            long_break_interval: 4,
        }
    }
}

/// Snapshot of the timer state sent to the frontend.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PomodoroSnapshot {
    pub phase: String,
    pub paused: bool,
    pub time_remaining_sec: u64,
    pub time_elapsed_sec: u64,
    pub current_tag: Option<String>,
    pub pomodoros_until_long_break: u32,
    pub sessions: Vec<Session>,
    pub settings: Settings,
    pub cycle_count: u32,
}

/// Full Pomodoro timer state.
pub struct PomodoroState {
    pub phase: Phase,
    pub paused: bool,
    pub time_remaining_sec: u64,
    pub time_elapsed_sec: u64,
    pub current_tag: Option<String>,
    pub cycle_count: u32, // pomodoros completed in current cycle (resets after long break)
    pub sessions: Vec<Session>,
    pub settings: Settings,
    pub data_path: PathBuf,
    /// Set when a working session starts, used to record started_at.
    pub session_started_at: Option<DateTime<Utc>>,
}

impl PomodoroState {
    pub fn new(data_path: PathBuf) -> Self {
        let (sessions, settings) = load_sessions(&data_path);
        Self {
            phase: Phase::Idle,
            paused: false,
            time_remaining_sec: 0,
            time_elapsed_sec: 0,
            current_tag: None,
            cycle_count: 0,
            sessions,
            settings,
            data_path,
            session_started_at: None,
        }
    }

    pub fn snapshot(&self) -> PomodoroSnapshot {
        let remaining_in_cycle = self.settings.long_break_interval - (self.cycle_count % self.settings.long_break_interval);
        PomodoroSnapshot {
            phase: self.phase.as_str().to_string(),
            paused: self.paused,
            time_remaining_sec: self.time_remaining_sec,
            time_elapsed_sec: self.time_elapsed_sec,
            current_tag: self.current_tag.clone(),
            pomodoros_until_long_break: remaining_in_cycle,
            sessions: self.sessions.clone(),
            settings: self.settings.clone(),
            cycle_count: self.cycle_count,
        }
    }

    /// Start a new pomodoro work session.
    pub fn start(&mut self, tag: Option<String>) {
        if self.phase != Phase::Idle {
            return;
        }
        self.phase = Phase::Working;
        self.paused = false;
        self.time_remaining_sec = self.settings.work_duration_sec;
        self.time_elapsed_sec = 0;
        self.current_tag = tag;
        self.session_started_at = Some(Utc::now());
    }

    /// Pause the timer.
    pub fn pause(&mut self) {
        if self.phase == Phase::Idle || self.paused {
            return;
        }
        self.paused = true;
    }

    /// Resume the timer.
    pub fn resume(&mut self) {
        if !self.paused {
            return;
        }
        self.paused = false;
    }

    /// Skip to next phase.
    pub fn skip(&mut self) {
        match self.phase {
            Phase::Idle => {}
            Phase::Working => {
                // Record the session as completed even on skip
                self.record_session(true);
                self.transition_after_work();
            }
            Phase::ShortBreak | Phase::LongBreak => {
                self.phase = Phase::Idle;
                self.paused = false;
                self.time_remaining_sec = 0;
                self.time_elapsed_sec = 0;
                self.current_tag = None;
            }
        }
    }

    /// Stop and return to idle (abandon current session).
    pub fn stop(&mut self) {
        if self.phase == Phase::Idle {
            return;
        }
        if self.phase == Phase::Working {
            self.record_session(false);
        }
        self.phase = Phase::Idle;
        self.paused = false;
        self.time_remaining_sec = 0;
        self.time_elapsed_sec = 0;
        self.current_tag = None;
        self.session_started_at = None;
    }

    /// Set/change the current tag.
    pub fn tag(&mut self, label: String) {
        self.current_tag = Some(label);
    }

    /// Tag a completed session by id.
    pub fn tag_session(&mut self, session_id: &str, label: String) {
        if let Some(s) = self.sessions.iter_mut().find(|s| s.id == session_id) {
            s.tag = label;
            self.save_sessions();
        }
    }

    /// Delete a session by id.
    pub fn delete_session(&mut self, session_id: &str) {
        self.sessions.retain(|s| s.id != session_id);
        self.save_sessions();
    }

    /// Tick the timer by 1 second. Returns true if a phase transition happened.
    pub fn tick(&mut self) -> bool {
        if self.phase == Phase::Idle || self.paused {
            return false;
        }

        if self.time_remaining_sec == 0 {
            return false;
        }

        self.time_remaining_sec -= 1;
        self.time_elapsed_sec += 1;

        if self.time_remaining_sec == 0 {
            match self.phase {
                Phase::Working => {
                    self.record_session(true);
                    self.transition_after_work();
                    return true;
                }
                Phase::ShortBreak | Phase::LongBreak => {
                    self.phase = Phase::Idle;
                    self.paused = false;
                    self.time_remaining_sec = 0;
                    self.time_elapsed_sec = 0;
                    self.current_tag = None;
                    return true;
                }
                Phase::Idle => {}
            }
        }

        false
    }

    fn transition_after_work(&mut self) {
        self.cycle_count += 1;
        if self.cycle_count % self.settings.long_break_interval == 0 {
            self.phase = Phase::LongBreak;
            self.time_remaining_sec = self.settings.long_break_sec;
        } else {
            self.phase = Phase::ShortBreak;
            self.time_remaining_sec = self.settings.short_break_sec;
        }
        self.time_elapsed_sec = 0;
        self.paused = false;
        self.current_tag = None;
        self.session_started_at = None;
    }

    fn record_session(&mut self, completed: bool) {
        let now = Utc::now();
        let started_at = self.session_started_at.unwrap_or(now);
        let elapsed = self.time_elapsed_sec;

        let next_id = self.next_session_id();
        let session = Session {
            id: next_id,
            tag: self.current_tag.clone().unwrap_or_else(|| "untitled".to_string()),
            category: "work".to_string(),
            started_at: started_at.to_rfc3339_opts(chrono::SecondsFormat::Secs, true),
            ended_at: Some(now.to_rfc3339_opts(chrono::SecondsFormat::Secs, true)),
            duration_sec: elapsed,
            completed,
        };
        self.sessions.push(session);
        self.save_sessions();
    }

    fn next_session_id(&self) -> String {
        let max = self
            .sessions
            .iter()
            .filter_map(|s| s.id.strip_prefix("s-").and_then(|n| n.parse::<u32>().ok()))
            .max()
            .unwrap_or(0);
        format!("s-{}", max + 1)
    }

    fn save_sessions(&self) {
        if let Some(parent) = self.data_path.parent() {
            let _ = fs::create_dir_all(parent);
        }
        let file = SessionFile {
            sessions: self.sessions.clone(),
            settings: self.settings.clone(),
        };
        let json = serde_json::to_string_pretty(&file).expect("serialize sessions");
        let _ = fs::write(&self.data_path, json);
    }

    /// Compute today's stats.
    pub fn today_stats(&self) -> (u32, u64) {
        let today = Utc::now().date_naive();
        let mut count = 0u32;
        let mut total_sec = 0u64;
        for s in &self.sessions {
            if s.completed {
                if let Ok(dt) = DateTime::parse_from_rfc3339(&s.started_at) {
                    if dt.date_naive() == today {
                        count += 1;
                        total_sec += s.duration_sec;
                    }
                }
            }
        }
        (count, total_sec)
    }
}

fn load_sessions(path: &PathBuf) -> (Vec<Session>, Settings) {
    if !path.exists() {
        return (Vec::new(), Settings::default());
    }
    match fs::read_to_string(path) {
        Ok(data) => match serde_json::from_str::<SessionFile>(&data) {
            Ok(file) => (file.sessions, file.settings),
            Err(_) => (Vec::new(), Settings::default()),
        },
        Err(_) => (Vec::new(), Settings::default()),
    }
}

/// Seed from seed.json if data file doesn't exist.
pub fn seed_if_needed(data_path: &PathBuf, seed_path: &PathBuf) {
    if data_path.exists() {
        return;
    }
    if seed_path.exists() {
        if let Some(parent) = data_path.parent() {
            let _ = fs::create_dir_all(parent);
        }
        let _ = fs::copy(seed_path, data_path);
    }
}

/// Compute salience for a session based on how long ago it was completed.
pub fn session_salience(session: &Session) -> (f64, Option<String>) {
    let now = Utc::now();
    if let Some(ended) = &session.ended_at {
        if let Ok(dt) = DateTime::parse_from_rfc3339(ended) {
            let mins = (now - dt.with_timezone(&Utc)).num_minutes();
            if mins < 60 {
                return (0.6, Some(format!("Completed {} min ago", mins)));
            } else if mins < 180 {
                let hours = mins / 60;
                return (0.4, Some(format!("Completed {}h ago", hours)));
            } else {
                let hours = mins / 60;
                return (0.2, Some(format!("Completed {}h ago", hours)));
            }
        }
    }
    (0.2, None)
}

/// Format seconds as MM:SS.
pub fn format_time(secs: u64) -> String {
    let m = secs / 60;
    let s = secs % 60;
    format!("{:02}:{:02}", m, s)
}
